File: app/src/main/java/me/instabattle/app/activities/BattleListActivity.java
package me.instabattle.app.activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.ListView;

import java.util.List;

import me.instabattle.app.managers.BattleManager;
import me.instabattle.app.R;
import me.instabattle.app.adapters.BattleListAdapter;
import me.instabattle.app.models.Battle;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class BattleListActivity extends Activity {

    private static final String TAG = "BattleListActivity";

    private ListView battleList;
    private BattleListAdapter battleListAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_battle_list);

        battleList = (ListView) findViewById(R.id.battleList);

        BattleManager.getAllBattlesAndDo(new Callback<List<Battle>>() {
            @Override
            public void onResponse(Call<List<Battle>> call, Response<List<Battle>> response) {
                battleListAdapter = new BattleListAdapter(BattleListActivity.this, response.body());
                battleList.setAdapter(battleListAdapter);
            }

            @Override
            public void onFailure(Call<List<Battle>> call, Throwable t) {
                //TODO
                Log.e(TAG, "cant get battles:" + t);
            }
        });
    }

    @Override
    public void onBackPressed() {
        Intent voting = new Intent(this, MenuActivity.class);
        startActivity(voting);
    }
}

File: app/src/main/kotlin/me/instabattle/app/activities/BattleListActivity.kt
package me.instabattle.app.activities

import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import me.instabattle.app.R
import me.instabattle.app.adapters.BattleListAdapter
import me.instabattle.app.managers.BattleManager
import me.instabattle.app.models.Battle
import org.jetbrains.anko.error
import org.jetbrains.anko.toast
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class BattleListActivity : DefaultActivity() {
    private lateinit var recyclerView: RecyclerView
    private lateinit var viewAdapter: RecyclerView.Adapter<*>
    private lateinit var viewManager: RecyclerView.LayoutManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_battle_list)

        viewManager = LinearLayoutManager(this)

        BattleManager.getAllAndDo(object : Callback<List<Battle>> {
            override fun onResponse(call: Call<List<Battle>>, response: Response<List<Battle>>) {
                viewAdapter = BattleListAdapter(this@BattleListActivity, response.body()!!)

                recyclerView = findViewById<RecyclerView>(R.id.battleList).apply {
                    setHasFixedSize(true)
                    layoutManager = viewManager
                    adapter = viewAdapter
                }
            }

            override fun onFailure(call: Call<List<Battle>>, t: Throwable) {
                toast("Failed to get battles, try again later.")
                error {"cant get battles:$t"}
            }
        })
    }

    override fun onResume() {
        super.onResume()

        if (::viewAdapter.isInitialized)
            viewAdapter.notifyDataSetChanged()
    }
}
